Cart discount code

katmatt opened the pull request:
Closes #674

katmatt commented:
@lauraluiz May be we should add an example cart discount with some discount codes to https://github.com/commercetools/commercetools-sunrise-data ? Impex unfortunately doesn't support `CartDiscount`, which means that we have to create our own json import for them 😞 WDYT?

lauraluiz commented on shopping-cart/pt/com/commercetools/sunrise/shoppingcart/adddiscountcode/DefaultAddDiscountCodeFormDataTest.java:
```diff
@@ -0,0 +1,44 @@
+                form.bind(addDiscountCodeFormData(""));
+        assertThat(validFormData.errors()).hasSize(1);
+        assertThat(validFormData.error("code")).isNotNull();
+    }
```
I'm missing here the case where the value is missing (not only empty).

lauraluiz commented on shopping-cart/test/com/commercetools/sunrise/shoppingcart/adddiscountcode/DefaultAddDiscountCodeControllerActionTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Before
+    public void setup() {
+        MockitoAnnotations.initMocks(this);
```
Do you like this option better than the class annotation?

katmatt replied:
No, I prefer to use the annotation 😉 Looks like I copied it from `DefaultResetPasswordControllerActionTest` and would suggest to change that test to use annotations too. WDYT?

lauraluiz replied:
That's what I thought, I wanted to ask you again just in case 😆 
Yep, we should change that test too then :)

lauraluiz commented on shopping-cart/test/com/commercetools/sunrise/shoppingcart/adddiscountcode/DefaultAddDiscountCodeControllerActionTest.java:
```diff
@@ -0,0 +1,71 @@
+            when(hookRunner.runActionHook(any(), any(), any()))
+                    .thenAnswer(invocation -> CompletableFuture.completedFuture(invocation.getArgument(2)));
+            when(sphereClient.execute(any()))
+                    .thenReturn(CompletableFuture.completedFuture(cart));
```
I'm missing the meaning of these mockings, I do have to fully understand them to guess what is the intention behind them. I would hide them behind meaningful method (mockSphereClientThatReturns...) and/or variable (dummy, fake) names. Something that can be read so that the purpose of the test becomes clear without having to figure out the code.

lauraluiz commented on shopping-cart/test/com/commercetools/sunrise/shoppingcart/removediscountcode/DefaultRemoveDiscountCodeControllerActionTest.java:
```diff
@@ -0,0 +1,73 @@
+            final UpdateAction<Cart> cartUpdateAction = cartUpdateCommand.getUpdateActions().get(0);
+            assertThat(cartUpdateAction).isInstanceOf(RemoveDiscountCode.class);
+            final RemoveDiscountCode removeDiscountCode = (RemoveDiscountCode) cartUpdateAction;
+            assertThat(removeDiscountCode.getDiscountCode().getId()).isEqualTo(discountCodeId);
```
That's super verbose, but I remember fighting to make it smaller and failed miserably because of the `List<? extend UpdateAction<Cart>>` :\

katmatt commented on shopping-cart/app/com/commercetools/sunrise/shoppingcart/adddiscountcode/DefaultAddDiscountCodeControllerAction.java:
```diff
@@ -0,0 +1,30 @@
+
+    protected CartUpdateCommand buildRequest(final Cart cart, final AddDiscountCodeFormData addDiscountCodeFormData) {
+        final AddDiscountCode addDiscountCode = AddDiscountCode.of(addDiscountCodeFormData.code());
+        return CartUpdateCommand.of(cart, Arrays.asList(addDiscountCode));
```
Good catch 👍

katmatt commented on shopping-cart/app/com/commercetools/sunrise/shoppingcart/adddiscountcode/SunriseAddDiscountCodeController.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public PageContent createPageContent(final Cart cart, final Form<? extends AddDiscountCodeFormData> form) {
+        final CartPageContent cartPageContent = pageContentFactory.create(cart);
+        cartPageContent.getCart().setAddDiscountCodeForm(form);
```
I didn't feel very comfortable with this code too 😞 I would then suggest to add another create method with an additional parameter, but I agree with you that it's a more general problem.

katmatt commented on it/controllers/shoppingcart/AddDiscountCodeControllerIntegrationTest.java:
```diff
@@ -0,0 +1,154 @@
+                        bind(SphereClient.class).toInstance(sphereClient);
+                        bind(CartInSession.class).toInstance(cartInSession);
+                    }
+                }).build();
```
That's much nicer 👍

katmatt commented:
@lauraluiz I've fixed all issues. Looking forward to your feedback 😄

katmatt commented:
@lauraluiz I fixed the failing tests and changed the description in the discount code view model to LocalizedString. Looking forward to your feedback/approval 😉
